### ISSUE TITLE
TextInput rendering certain characters as boxes

### ISSUE THREAD

**Issue by Leva7** [kivy/kivy#4741]
There is a certain set of Unicode characters that are rendered as boxes in the `TextInput`.  
What I mean is that although the characters are included in the font and don't seem to have much of a difference from other (letters and stuff like that), their texture is not shown in `TextInput` and `Label` widgets. The characters that were used in the test are emojis in range `0x1F600`-`0x1F61F`

Here are some screenshots to describe the problem  
This is how the text looks in LibreOffice (`.txt` file):
![libre](https://cloud.githubusercontent.com/assets/15035286/20456599/dee2c62e-aea3-11e6-8f78-fe689c4293a0.png)
And this is what a `TextInput` displays:
![kivy](https://cloud.githubusercontent.com/assets/15035286/20456601/ecac47f8-aea3-11e6-87f5-715a88e1e33c.png)

[Here](http://stackoverflow.com/questions/40002196/kivy-textinput-not-rendering-some-characters-in-the-font) is a question on Stack Overflow about this problem

OS: Ubuntu 16.04
Python: 3.5.2
Kivy: 1.9.2-dev0 (daily build)
Text provider: sdl2
Used font: NotoSans-Regular ([get it from Google](https://www.google.com/get/noto/#sans-lgc))

**Comment by KeyWeeUsr** [comment]
Finally! I was finally able to reproduce the same inserting of the emoji I did before with the same font you have.

> Samsung's default keyboard, kivy-master, android 4.4.2

the first up-left emoji character is displayed hm... not exactly as you'd expect, but as I [mentioned](http://stackoverflow.com/questions/40002196/kivy-textinput-not-rendering-some-characters-in-the-font#comment67318599_40002196). However I'm sorry to tell that the rest of emojis don't work for me with this font and are displayed only as the boxes above.

![image](https://cloud.githubusercontent.com/assets/8825439/20457186/c57bf03e-ae85-11e6-9ea8-223573d7d147.png)

**Comment by Leva7** [comment]
@KeyWeeUsr no, this character is displayed as expected. In Unicode, there are two emojis (or more, idk) that are not part of the main emoji range (link [here](http://unicode.org/emoji/charts/full-emoji-list.html)). And apparently these characters are not in the "set" that I've mentioned so they are displayed both in your Android and in my PC.  
I assume this issue could have something to do with the glyph complexity

**Comment by tito** [comment]
Are you sure the glyph is contained in the font?
Cause unlike the Libreoffice or the OS, currently, we don't have glyph fallback at all. Either the glyph is contained in the font, either it's not. This is something we need to work on.

I tried to test with https://fonts.google.com/?query=noto on the glyph U+1F600 (😀), it appear as a box too on Google Font.
You can try to analyse the font more deeply to see if it exists
(or very hackish, make a webpage, include noto, use it as default, put your smiley somewhere, then inspect the element, go to the "computed styles", at the bottom, you'll see the real font used for the glyph (if you write a sentence, it will show how many glyph are rendered from this font, and how many from anothers.)

Kivy itself have no issue to render exotic glyph as long as they are included in the font :)

**Comment by Leva7** [comment]
I've checked it with High-Logic Font Creator, the glyph _definitely is_ in the .ttf font.
I don't know what have you seen on Google Font, as my browser displays both U+1F600 and U+1F601.
![image](https://cloud.githubusercontent.com/assets/15035286/21045899/5f9e4a3a-be2c-11e6-8db4-d77dd9c55c4a.png)

**Comment by udiboy1209** [comment]
@Leva7 check the character list here : https://fonts.google.com/specimen/Noto+Sans . It doesn't show your glyph.

**Comment by tito** [comment]
Yes, your glyph is not in the font. I download Noto Sans Regular.ttf, used http://torinak.com/font/lsfont.html
Import, see all the glyph in it, your smiley is not.

Glyph-fallback is something missing in Kivy

**Comment by Leva7** [comment]
@tito @udiboy1209 Here's the thing, should have mentioned earlier. I manually combined the Noto Emoji and Noto Sans Regular, so I see these glyphs. But even then, if I use the default non-modified Noto Emoji, it still doesn't show it

**Comment by KeyWeeUsr** [comment]
I see emojis (the grinning one mainly) in browser as `Segoe UI Emoji` or `EmojiOne` font. I tried to use the font (it's in `C:/Windows/Fonts`, I'm not sure about license so you need to rip it yourself from there) and also the `EmojiOne`, but guess what...

Yeah, nothing. I was totally confused, because I put in python directly `u"\U0001F600"`, but then I opened the files with http://torinak.com/font/lsfont.html

Android Emoji - http://forum.xda-developers.com/attachment.php?attachmentid=1315721&d=1347277568 totally blank

Emoji One - https://github.com/eosrei/emojione-color-font/releases/download/v1.3/EmojiOneColor-SVGinOT-1.3-Win.zip the same and also for Noto fonts, which you can see in the upper part of my image.

If however such images were available **in the font file**, then you'd see the bottom part of my image:

![image](https://cloud.githubusercontent.com/assets/8825439/21049804/3678d6ee-be17-11e6-8dcc-c21b0a85b122.png)

**Comment by Leva7** [comment]
@tito @KeyWeeUsr I think there's some issue with the way Kivy and torinak.com/font/lsfont.html read the font. When I've loaded the [NotoEmoji-Regular.ttf](https://www.google.com/get/noto/#emoji-zsye) in there, I could only see glyphs up to 0xFFFF, even though the Unicode range is greater and there clearly are more glyphs outside 0x0000 - 0xFFFF. Is that a possibility?

@udiboy1209 I don't see a character list in the link you provided, could you show me please?

**Comment by KeyWeeUsr** [comment]
@Leva7 Ding ding, you found the problem!

@tito http://www.filedropper.com/osetest I put some emoji from the file into a separate test `ttf` font. There are emojis for range `a-z`, which are rendered correctly as b&w. Which triggers a question - why the range `>FFFF` isn't supported / renders boxes?

**Comment by dessant** [comment]
I don't see the emoji either with this example:

```python
from kivy.app import App
from kivy.lang import Builder

kv = """
GridLayout:
    cols: 1
    Label:
        font_name: 'NotoEmoji-Regular.ttf'
        text: '\U0001F600'
    TextInput:
        font_name: 'NotoEmoji-Regular.ttf'
        text: '\U0001F600'
"""


class TestApp(App):
    def build(self):
        return Builder.load_string(kv)

if __name__ == '__main__':
    TestApp().run()
```

**Comment by goffi-contrib** [comment]
Hello, I have the same issue with a Label (tested on 1.9.1 and git). The following code reproduce the issue:

```
#!/usr/bin/env python2
#-*- coding: utf-8 -*-

from kivy.app import App
from kivy.lang import Builder

root = Builder.load_string(u'''
Label:
    text: u"😉"
    font_size: '64pt'
    font_name: "/usr/share/fonts/noto/NotoEmoji-Regular.ttf"
''')


class Label(App):
    def build(self):
        return root

if __name__ == '__main__':
    Label().run()
```

After spending a lot of time on it, I finally found that it is due to this bug: https://github.com/kivy/kivy/issues/2761

I've checked, the bug is still here, and if you do `export KIVY_TEXT=pil` the smiley is shown as expected.

I think it would be interesting to consider integrating Pango, this would solve this issue, handle fallbacks and multilingual texts. Actually it seems that some work has already be done in this direction: https://github.com/kivy/kivy/issues/2669#issuecomment-64964386

@kived is the Pango rendered still in progress or is it abandoned? Thanks

**Comment by KeyWeeUsr** [comment]
@goffi-contrib I don't see it with master and PIL. Not the exact copied emoji, nor unicode string in the example above. Could you test with 1.9.2?

**Comment by goffi-contrib** [comment]
hello @KeyWeeUsr . I've tested again on the last git (tested on Arch with aur/python2-kivy-git), and I can confirm it's still the same:  a square with default text renderer (sdl2), while the emoji appears correctly with PIL (using `export KIVY_TEXT=pil`).

I really think integrating Pango is best way to fix that and other text issues.

**Comment by dessant** [comment]
I can't make it work with the Pillow text provider either. Windows 10, Python 3.4.4, Kivy 57d41c9, Pillow 4.0.0.

**Comment by KeyWeeUsr** [comment]
Same here with Python 3.5.2 and Win 8.1. I don't think using the old PIL (e.g. 1.1.7) instead of Pillow would help here. @goffi-contrib could you test on Windows?

**Comment by goffi-contrib** [comment]
Hi, no I can't sorry, I have no Windows machine available.

**Comment by jd-0001** [comment]
It's been 6 years and can't we render the emoji in kivy 🤯

**Comment by Sahil-pixel** [comment]
is this solved or not ?

**Comment by kengoon** [comment]
with the latest version of kivy(master branch 2.2.0.dev0), I can confirm that kivy actually renders colored emojis
![Screenshot from 2022-07-03 10-36-44](https://user-images.githubusercontent.com/42192162/177034009-1ed9e82a-dc0c-4711-a6c4-833249de97c7.png)

**Comment by kengoon** [comment]
> with the latest version of kivy(master branch 2.2.0.dev0), I can confirm that kivy actually renders colored emojis using `sdl2` as the text provider.
> 
> code
> 
> ```python
> from kivy.app import App
> from kivy.lang import Builder
> 
> root = Builder.load_string(u'''
> Label:
>     text: "🔞"
>     font_size: '64pt'
>     font_name: "/usr/share/fonts/truetype/noto/NotoColorEmoji.ttf"
> ''')
> 
> 
> class Label(App):
>     def build(self):
>         return root
> 
> if __name__ == '__main__':
>     Label().run()
> ```
> 
> ![Screenshot from 2022-07-03 10-36-44](https://user-images.githubusercontent.com/42192162/177034009-1ed9e82a-dc0c-4711-a6c4-833249de97c7.png)

there is an issue, the font_size does not respond

**Comment by RobinPicard** [comment]
I've opened a PR to fix the issue with the color of emojis in `TextInput` widgets: https://github.com/kivy/kivy/pull/7997

**Comment by jd-solanki** [comment]
> I've opened a PR to fix the issue with the color of emojis in `TextInput` widgets: #7997

oh, thanks so much for this ❤️

**Comment by RobinPicard** [comment]
> there are issues though, the font_size does not respond and the text does not render with color on `TextInput` widget

I think the font size issue is unrelated. I've tried with a different emoji font than NotoColorEmoji and things worked fine. The problem comes from the values returned by the `TTF_OpenFont` function with this specific font. I guess there's something wrong with either the font or SDL_ttf

**Comment by RobinPicard** [comment]
I tried with Noto Emoji: https://fonts.google.com/noto/specimen/Noto+Emoji

**Comment by kengoon** [comment]
It matters because it is not colored

**Comment by ElliotGarbus** [comment]
@Julian-O should the status be changed to has PR?
see: #7997
